### ISSUE TITLE
Conversations is very slow to display/open messages

### ISSUE THREAD

**Issue by sbernard31** [thunderbird-conversations/thunderbird-conversations#1620]
* **Thunderbird version:** 78.8
* **Thunderbird Conversation version:** 3.2.16
* **Operating System:** Debian 10 Stable (buster)

**What did you do?**
Click on a message in messages list.

**What did you expect to happen?**
The message is displayed quickly in message view.

**What actually happened?**
Nothing happens during several seconds + CPU to 100%, then finally the message is displayed as expected.

**Comment by taylormia** [comment]
Hello, I'm on TB 88.0b1 on Manjaro Linux 21.0 on a Dell XPS 9560 and Conversations v. 3.2.1.6 and I'm experiencing very similar slowness in displaying the thread of a message. The same slowness happens in TB 87 and 86. When I click on a message it takes about 20 seconds to display the thread in the reading pane. The CPU does not spike to 100% in my case. 
I love this add-on, but productivity dies when you have to wait 20 sec to display a conversation thread.
Thanks for an suggestions to speed up the thread display.

**Comment by Standard8** [comment]
@taylormia Please can you try getting a profile as detailed here: https://github.com/thunderbird-conversations/thunderbird-conversations/wiki/Profiling-Conversation's-Performance

**Comment by Standard8** [comment]
It'd also be useful to know how many messages are displayed once you've clicked on the one message - is it just one or a few, are there lots of contacts?

**Comment by taylormia** [comment]
As an example, when there are two messages in the conversation thread - it takes about 5-7 seconds. When there are 12-15 messages it takes about 20 seconds. My contacts are coming from an Office 365 corporate Global Address Book - so there are thousands. 
I will get you a performance profile in a hour or two...stay tuned

**Comment by taylormia** [comment]
> @taylormia Please can you try getting a profile as detailed here: https://github.com/thunderbird-conversations/thunderbird-conversations/wiki/Profiling-Conversation's-Performance

Here's the performance profile. Started recording when I clicked on a message with around 25-30 messages in the thread...and stopped recording when the messages appeared in 25-30 sec,
https://share.firefox.dev/3uilAlI

**Comment by Standard8** [comment]
@taylormia I've seen the original link and you should be able to delete it if you visit https://profiler.firefox.com

Can you file an [issue on the profiler tool](https://github.com/firefox-devtools/profiler/issues) with details of the info that wasn't stripped? I'll try and make sure the right people see it and hopefully get it fixed.

With regards to the profile itself, I can see that most of it nothing is happening - approximately every 1.5s we're starting a request to get a contact. So I assuming it is taking 1.5s to return contacts.

The large address book that you have, how is that loaded into Thunderbird - is it via an add-on, LDAP or some other import?

**Comment by Standard8** [comment]
@sbernard31 might you be in the same situation with respect to address books? e.g. a large one or one that is from a remote server?

**Comment by taylormia** [comment]
@Standard8 My contacts are via LDAP connecting to a remote Active Directory address book. I use DavMail as an EWS gateway to Office 365 and this allows connection to a remote LDAP directory for contacts.

**Comment by taylormia** [comment]
@Standard8 Can the add-on check the downloaded offline copy of the LDAP directory instead of the online directory? I notice that when I put TB into Offline Mode, the messages in the thread are shown almost instantaneously.

**Comment by taylormia** [comment]
@Standard8 Do you believe the "Move the Contacts cache into background.js" change will fix the slowness issue? If so, when do you expect to have it ready for testing?

**Comment by Standard8** [comment]
> @Standard8 Can the add-on check the local downloaded offline copy of the LDAP directory instead of the online directory? I notice that when I put TB into Offline Mode, the messages in the thread are shown almost instantaneously.

From the add-on perspective we only have what Thunderbird provides us - which is the online directory, we can't get into the offline data unless Thunderbird gives it to us. This would likely need something like [automatic sync](https://bugzilla.mozilla.org/show_bug.cgi?id=532684) on the Thunderbird side, but I'm fairly sure there hasn't been much attention paid to the LDAP side of Thunderbird for a few years now unfortunately.

> FYI..In normal usage of the LDAP address book..there are no delays..address lookups take 1-2 seconds.

The issue is mainly because we're doing sequential lookups of every different email address referenced in messages in the thread - there is some caching but that is only per conversation display.

> @Standard8 Do you believe the "Move the Contacts cache into background.js" change will fix the slowness issue? If so, when do you expect to have it ready for testing?

It will probably help a little bit, and work towards helping some more. The primary change it will make is to cache contacts across the whole session, so some of the messages may be slow especially when starting, but you should find that it improves as you go through your session. I think we'll probably be ready to get that out in a week or two.

We are discussing a follow-up change, which will make the contacts lookup happen in parallel to the display of emails - so we'll focus on displaying the messages first, and the contact lookups happen (if we haven't got them cached already) in the background and the display will be updated once the details are received.

**Comment by micker** [comment]
for me same problem ... if i disable plugin all seams good ...
i try to change option to doesn't download email on disk for imap but nothing more...
how to solve it ?
win10 th 78.9.1

**Comment by taylormia** [comment]
@standard Very much looking forward to the first fix in a week or so...and also the "parallel contacts lookup" change later. Please keep us posted. Thanks

**Comment by micker** [comment]
somebody have a solution ?
regards

**Comment by Standard8** [comment]
@micker Sorry, we're not always active all the time. From the profile I'm not quite sure what is up, is that a really large message thread you are loading, or maybe one with lots of contacts?

It also looks like there's indexing of your database going on in the background. If you look at tools -> activity manager you should be able see if that is still going on or not (it can take a couple of hours), and maybe make sure that looks to have finished before loading a thread.

**Comment by micker** [comment]
Hello sorry i understand... It better last day but now lagg is return ... How i Can help more ?

**Comment by Standard8** [comment]
> Hello sorry i understand... It better last day but now lagg is return ... How i Can help more ?

Is this happening on just large message threads or ones with lots of contacts, or does it happen even on single messages?

Did you check to see if the activity manager had finished?

**Comment by Standard8** [comment]
Good news, 3.2.21 was published earlier today. This includes the contact caching in the background. The contact caching reduces look-ups in the address book for where we are checking for the same email address multiple times when loading a message thread. It also caches and saves the email addresses (in memory), for subsequent thread loads.

Whilst you may still see slow loading due to the contacts lookup initially, once the cache is populated, then it should become faster.

We do hope in future to make the contact lookup asynchronous to the message display, but currently that requires more rearchitecture work as we don't currently have the contact information soon enough to make it worth it.

**Comment by taylormia** [comment]
Does the contact caching look at all the local and remote LDAP address books? Does it cache all the addresses in the remote address book? In my case, I use a corporate LDAP address book with about 10,000 contacts. Does it cache all these contacts?
I ask because I have now enabled the add-on and it has been over 5 min....and the message has not displayed yet int he reading pane.

**Comment by Standard8** [comment]
Make sure you are on definitely 3.2.21 - it might not have updated yet. There's a cog symbol in the add-on manager to check for updates.

The cache only caches contacts that are on emails you read, it does not load the whole address book and then cache it.

It will cache up to 1000 contacts, and then drop the oldest 250 when it breaks that limit.

My guess is that LDAP is going to be quite slow at least initially as you've round trips to the server. If you're loading a big message thread with lots of contacts it might still take a while until it has cached them, but switching away and back to that thread should be a lot quicker.

If it isn't helping, knowing the stats of what you're loading, and getting another profile might help.

**Comment by taylormia** [comment]
I am running 3.2.21. Unfortunately this is unusable for me. For example I clicked on a thread containing 10 messages with 4 unique contacts. All four contacts are frequently used contacts, so they are available in the Collected Addresses address book locally. It has been over 8 min now and the thread has yet to display in the message pane. Just for reference - looking up a contact from the remote LDAP address book using TB takes a couple of seconds. I'm using TB 90.0b1

**Comment by Standard8** [comment]
Please could you try going to Tools -> Developer Tools -> Error Console and see if you are getting any error messages appearing when you select the messages. I'm wondering if there's something happening in the background that is causing an error.

Also, could you try it in offline mode? (or maybe temporarily remove the LDAP address book)? I want to eliminate the potential for LDAP to be an issue here.

**Comment by taylormia** [comment]
When I'm online and click on a thread, I get these errors
![image](https://user-images.githubusercontent.com/34145387/122261777-57e76500-ce9a-11eb-8166-a7c06051945d.png)

I can access the remote and local address books without an issue and the response is almost instantaneous

**Comment by micker** [comment]
for me no change after update and th 78. doesnt i need to waiting and keep th open for creating cache ?

**Comment by Standard8** [comment]
@micker I think yours is a different issue, I think you need to file a new issue and answer the questions I asked previously.

**Comment by Standard8** [comment]
@taylormia thank you for the information. I've tracked this down to an issue with the Thunderbird betas and the LDAP directory. The Thunderbird betas have enabled add-ons to search in LDAP address books, unfortunately, if two searches are kicked off at the same time, then one of them may fail - unfortunately doing searches in parallel is one way I tried to make Conversations faster.

The error you saw is also something to do with LDAP not working correctly when it does find contacts in those address books.

I've filed bugs on Thunderbird for both of [these](https://bugzilla.mozilla.org/show_bug.cgi?id=1716858) [issues](https://bugzilla.mozilla.org/show_bug.cgi?id=1716861). I'm in contact with the maintainers, so we will hopefully get those looked at in the next day or so.

Once I hear back, I'll work out if I need to do some sort of modification to handle LDAP in the slower fashion, or if we can wait for Thunderbird to fix things.

**Comment by taylormia** [comment]
Thanks for tracking this down. Looking forward to see if there will be a fix either from TB or your side or both.

**Comment by taylormia** [comment]
I recall using an earlier version of Conversations with an earlier version of TB that did not have the delay issues. Was there a version of Conversations in which you changed the code which started causing the delays? if so,  I'd like to try the earlier version to see if the delays are there with TB 90.b2. Any thoughts?

**Comment by Standard8** [comment]
@taylormia in the most recent update (3.2.24) I changed the lookup to not do contact searches in parallel. For non-LDAP this doesn't affect performance by any easily measurable quantity, for LDAP this will of course be slower, but it should at least work.

There is a small risk you could still get into the "stuck" state, if you change between messages in the message list quickly - I can't really do much about that, we need the [Thunderbird issue fixed](https://bugzilla.mozilla.org/show_bug.cgi?id=1716861) - I have flagged that for the next release, so hopefully we'll get it fixed (or exclude LDAP).

**Comment by taylormia** [comment]
Version 3.2.24 is definitely better in terms of how long a thread takes to display - but not instant like it used to be. I'm going to keep it enabled and see if I can tolerate the 5-15 sec. delay in real life scenarios. Thanks for working on this and hopefully it will be fixed soon.

**Comment by taylormia** [comment]
On further testing, some threads are taking up to 45 sec to display. I cannot use TB effectively in this scenario - so, I'm going to disable it for now until this is fixed. Again, I appreciate all the work you are putting into this. I love the functionality of this add-on but it is not usable with these delays.

**Comment by micker** [comment]
form me same ... but maybe an idea i habe more delay when i have many replay in same discussion with many image inside email

**Comment by Standard8** [comment]
@taylormia 3.2.28 has just been released, make sure you're on 91.0b3 and that should fix the slowness for you.

Conversations won't look up contacts within any LDAP address books for now, I've filed #1689 to fix the architectural issues around that later.

**Comment by taylormia** [comment]
@Standard8 This works perfectly on 91.0b3. Display of threads is instant...just like it used to be. Yay !! Awesome job and thanks so much for fixing my most favorite TB add-on!!

**Comment by Standard8** [comment]
Thank you for confirming that, I'm glad that is much better now.

Also just to note there are a few known issues with 91betas that I'm working towards fixing, 3.2.28 fixes some of those, others are listed in #1686.

I'm going to close this out, although I'm not sure this was the original authors problem, there have been a few things fixed with respect to performance improvements, so hopefully this is better now. There's still bottlenecks that we'll look to reduce as we move forward though.

Other issues with performance should be filed [along with profiles](https://github.com/thunderbird-conversations/thunderbird-conversations/wiki/Profiling-Conversation's-Performance) as new issues.

**Comment by taylormia** [comment]
@Standard8 Just wanted to ask if you got my email that I sent to your personal email address
